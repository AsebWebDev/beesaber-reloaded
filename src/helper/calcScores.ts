import type { Scores } from '@/../sharedTypes';

const calcTopScores = (scoresRecent: Scores): Scores =>
  [...scoresRecent].sort((a, b) => b.score - a.score);

const calcScoreHashed = (scoresRecent: Scores): string[] => {
  const scoredSongsHashes: string[] = [];

  scoresRecent.forEach((element) => scoredSongsHashes.push(element.songHash));

  return scoredSongsHashes;
};

export { calcScoreHashed, calcTopScores };
