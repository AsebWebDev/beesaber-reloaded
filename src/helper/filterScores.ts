import type { Score, Scores } from '@/../sharedTypes';

const isInQuery = (score: Score, query: string): boolean => {
  if (query === '') return true;

  // check for string only containing blank spaces
  if (query.replace(/\s/g, '').length === 0) return false;

  const { songName, songAuthorName } = score;

  return (
    songName.toLowerCase().includes(query.toLowerCase()) ||
    songAuthorName.toLowerCase().includes(query.toLowerCase())
  );
};

const filterScores = (
  scores: Scores,
  query: string,
  filterByIsPlayedByHive: boolean
): Scores =>
  scores.filter((score) =>
    filterByIsPlayedByHive
      ? isInQuery(score, query) && score.playedByHive !== undefined
        ? score.playedByHive
        : false
      : isInQuery(score, query)
  );

export { filterScores, isInQuery };

export default filterScores;
