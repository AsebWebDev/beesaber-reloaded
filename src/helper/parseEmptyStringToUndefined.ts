export default (id: string | undefined): string | undefined =>
  id === '' ? undefined : id;
