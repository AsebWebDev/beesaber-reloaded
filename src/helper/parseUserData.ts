import type {
  Score,
  ScoreData,
  ScoreSaberUserInfo,
  UserData,
  UserScores,
} from '../../sharedTypes';

const parseUserData = (
  userData: UserData,
  scoreSaberData: ScoreSaberUserInfo,
  { scores: scoresRecent }: UserScores
): UserData => {
  const {
    playerInfo: { avatar: profilePic, playerName: username, rank },
    scoreStats,
  } = scoreSaberData;

  const scoresTop = [...scoresRecent].sort((a, b) => b.score - a.score); // sort descending
  const scoredSongsHashes = scoresRecent.map(
    (element: Score) => element.songHash
  );

  const parsedScoreData: ScoreData = {
    scoredSongsHashes,
    scoresRecent,
    scoresTop,
  };

  const parsedData = {
    ...userData,
    profilePic,
    rank,
    scoreData: parsedScoreData,
    scoreStats,
    username,
  };

  return parsedData;
};

export default parseUserData;
