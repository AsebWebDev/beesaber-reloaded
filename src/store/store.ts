import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/dist/query/react';

import { apiInitialState } from '@/api/api';
import { apiAuth } from '@/api/services/apiAuth/apiAuth';
import { apiPlayer } from '@/api/services/apiPlayer/apiPlayer';
import { apiUser } from '@/api/services/apiUser/apiUser';

import { initialState as appStatusInitialState } from './reducer/appStatusReducer';
import { initialState as notificationsInitialState } from './reducer/notificationsReducer';
import rootReducer from './reducer/rootReducer';
import { initialState as userDataInitialState } from './reducer/userDataReducer';

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(
      apiAuth.middleware,
      apiPlayer.middleware,
      apiUser.middleware
    ),
  devTools: true,
});

setupListeners(store.dispatch);

const initialState = {
  appStatus: { ...appStatusInitialState },
  notifications: { ...notificationsInitialState },
  userData: { ...userDataInitialState },
  api: { ...apiInitialState },
};

export { initialState };

export default store;

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
