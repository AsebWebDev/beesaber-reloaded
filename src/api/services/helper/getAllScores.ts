/* eslint-disable no-console */
import { apiPlayer } from '@/api/services/apiPlayer/apiPlayer';
import { calcScoreHashed, calcTopScores } from '@/helper/calcScores';
import store from '@/store/store';

import type { ScoreData, Scores } from '@/sharedTypes';

type Props = {
  array?: Scores;
  count?: number;
  id: string;
  threshold?: number;
};

/**
 * Fetch all recent Scores of a ScoreSaberUser
 * @param {String} id - ScoreSaber id
 * @param {number} count - first score page to fetch, default = 1
 * @param {Scores} array - initial array for recursive data fetching, default = []
 * @param {number} threshold - optional threshold for number of score pages (1 page = 8 scores)
 * @returns {Scores} Array of recent scores
 */
const getRecentScores = async ({
  id,
  count = 1,
  array = [],
  threshold,
}: Props): Promise<Scores> => {
  const dispatchedPromise = store.dispatch(
    apiPlayer.endpoints.getRecentScores.initiate({ id, count })
  );
  const { data: response } = await dispatchedPromise;

  if (response === undefined) return [];
  if (threshold !== undefined && count > threshold) return response;

  return array.concat(
    await getRecentScores({ id, count: count + 1, array: response, threshold })
  ) as Scores;
};

const getAllScores = async (
  id: string,
  threshold?: number
): Promise<ScoreData> => {
  const scoresRecent = await getRecentScores({ id, threshold });

  return {
    scoredSongsHashes: calcScoreHashed(scoresRecent),
    scoresRecent,
    scoresTop: calcTopScores(scoresRecent),
  };
};

export { getRecentScores };

export default getAllScores;
