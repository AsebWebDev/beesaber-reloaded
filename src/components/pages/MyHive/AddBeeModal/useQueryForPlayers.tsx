import { useEffect, useState } from 'react';

import {
  useGetFullPlayerQuery,
  useGetPlayersByNameQuery,
} from '@/api/services/apiPlayer/apiPlayer';
import { useGetUserDataQuery } from '@/api/services/apiUser/apiUser';
import useDebounce from '@/sharedHooks/useDebounce';
import { useAppSelector } from '@/store/hooks';
import { selectUserId } from '@/store/reducer/userDataReducer';

import type { PlayerInfo } from '@/sharedTypes/ScoreSaberUserInfo';

type Props = {
  query: string;
  searchBy: 'id' | 'name';
};

type ReturnType = {
  foundPlayers: PlayerInfo[] | null;
  showSpinner: boolean;
  thatIsYou: boolean;
  userAlreadyAdded: boolean;
};

const useQueryForPlayers = ({ query, searchBy }: Props): ReturnType => {
  const userId = useAppSelector(selectUserId);
  const { data: userData } = useGetUserDataQuery(userId);
  const [foundPlayers, setfoundPlayers] = useState<PlayerInfo[] | null>(null);
  const [thatIsYou, setThatIsYou] = useState(false);
  const [userAlreadyAdded, setUserAlreadyAdded] = useState(false);
  const debouncedSearchQuery = useDebounce(query, 600);

  const {
    data: playersByName,
    isFetching: isFetchingQueryByName,
    error: errorByName,
  } = useGetPlayersByNameQuery(debouncedSearchQuery);
  const {
    data: playerById,
    isFetching: isFetchingQueryById,
    error: errorById,
  } = useGetFullPlayerQuery(debouncedSearchQuery);

  const showSpinner = isFetchingQueryByName || isFetchingQueryById;

  useEffect(() => {
    if (
      foundPlayers !== null &&
      foundPlayers.length === 1 &&
      userData !== undefined &&
      userData.bees.length > 0
    ) {
      setThatIsYou(foundPlayers[0].playerId === userData.myScoreSaberId);
      setUserAlreadyAdded(
        userData.bees.some((item) => item.playerId === foundPlayers[0].playerId)
      );
    } else {
      setThatIsYou(false);
      setUserAlreadyAdded(false);
    }
  }, [foundPlayers, userData]);

  useEffect(() => {
    const searchHasError =
      (searchBy === 'id' && errorById !== undefined) ||
      (searchBy === 'name' && errorByName !== undefined);

    if (searchHasError) {
      setfoundPlayers(null);
    }
  }, [errorById, errorByName]);

  useEffect(() => {
    const players = searchBy === 'id' ? playerById?.playerInfo : playersByName;

    if (players === undefined) return;

    // If result is an array, multiple Users have been found...
    if (Array.isArray(players) && players.length > 1) setfoundPlayers(players);
    // If result is an array of 1 element, only one user has been found...
    if (Array.isArray(players) && players.length === 1)
      setfoundPlayers([players[0]]);
    // If result is not an array (searching by ID) only one user had been found...
    if (!Array.isArray(players)) setfoundPlayers([players]);
  }, [playerById, playersByName]);

  return { foundPlayers, showSpinner, thatIsYou, userAlreadyAdded };
};

export default useQueryForPlayers;
