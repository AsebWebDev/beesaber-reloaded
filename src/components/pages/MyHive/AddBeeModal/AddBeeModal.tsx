import { skipToken } from '@reduxjs/toolkit/dist/query/react';
import {
  MDBBtn,
  MDBContainer,
  MDBInput,
  MDBModal,
  MDBModalBody,
  MDBModalContent,
  MDBModalDialog,
  MDBModalFooter,
  MDBModalHeader,
  MDBTabs,
  MDBTabsContent,
  MDBTabsItem,
  MDBTabsLink,
  MDBTabsPane,
} from 'mdb-react-ui-kit';
import React, { useEffect, useState } from 'react';
import { toast } from 'react-toastify';
import styled from 'styled-components';

import {
  useGetUserDataQuery,
  useUpdateUserDataMutation,
} from '@/api/services/apiUser/apiUser';
import getAllScores from '@/api/services/helper/getAllScores';
import Alert from '@/components/common/Message/Message';
import NeonText from '@/components/common/NeonText/NeonText';
import Spinner from '@/components/common/Spinner/SpinnerPulse';
import { useAppSelector } from '@/store/hooks';
import { selectUserId } from '@/store/reducer/userDataReducer';
import tokens from '@/tokens';

import FoundPlayersList from './FoundPlayersList/FoundPlayersList';
import useQueryForPlayers from './useQueryForPlayers';

import type { Bee } from '@/../sharedTypes';
import type { PlayerInfo } from '@/../sharedTypes/ScoreSaberUserInfo';

const ModalContent = styled(MDBModalContent)`
  background-color: ${tokens.color.page.bgColor.light};
  color: ${tokens.color.white.main};
  margin-top: 10rem;
  width: fit-content;
`;

type Props = {
  toggleModal: () => void;
};

type Tab = '1' | '2';

const AddBeeModal = ({ toggleModal }: Props): JSX.Element | null => {
  const [activeItem, setActiveItem] = useState<Tab>('1');
  const [query, setQuery] = useState('');
  const [selectedPlayer, setSelectedPlayer] = useState<
    PlayerInfo | undefined
  >();
  const [beeToAdd, setBeeToAdd] = useState<Bee | undefined>();
  const searchBy = activeItem === '1' ? 'id' : 'name';
  const [updateUser] = useUpdateUserDataMutation();
  const userId = useAppSelector(selectUserId);
  const { data: userData } = useGetUserDataQuery(userId ?? skipToken);

  const { foundPlayers, showSpinner, thatIsYou, userAlreadyAdded } =
    useQueryForPlayers({ query, searchBy });

  useEffect(() => {
    if (foundPlayers !== null && foundPlayers.length === 1)
      setSelectedPlayer(foundPlayers[0]);
    else {
      setBeeToAdd(undefined);
      setSelectedPlayer(undefined);
    }
  }, [foundPlayers]);

  useEffect(() => {
    if (selectedPlayer === undefined) return;
    const { playerId, playerName } = selectedPlayer;
    const buildBee = async () => {
      const scoreData = await getAllScores(playerId);

      setBeeToAdd({
        playerId,
        playerName,
        scoreData,
      });
    };

    void buildBee();
  }, [selectedPlayer]);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setQuery(e.target.value);

  const handleSelect = (player: PlayerInfo) => {
    // select or toggle off player
    setSelectedPlayer(player === selectedPlayer ? undefined : player);
    setBeeToAdd(undefined);
  };

  const handleSave = async () => {
    if (
      userAlreadyAdded ||
      userData === undefined ||
      beeToAdd === undefined ||
      userId === undefined
    )
      return;

    await toast.promise(
      updateUser({
        userId,
        userData: { bees: [...userData.bees, beeToAdd] },
      }),
      {
        pending: `Saving ${beeToAdd.playerName}...`,
        success: `user ${beeToAdd.playerName} has been saved ðŸ‘Œ`,
        error: `There has been an issue saving ${beeToAdd.playerName} ðŸ¤¯`,
      }
    );
    setQuery('');
  };

  const switchTab = (tab: Tab) => {
    setQuery('');
    if (activeItem !== tab) setActiveItem(tab);
  };

  return (
    <MDBContainer>
      <MDBModal staticBackdrop show={true}>
        <MDBModalDialog>
          <ModalContent>
            <MDBModalHeader>Add a new Bee</MDBModalHeader>
            <MDBModalBody>
              <MDBTabs className="mb-3">
                <MDBTabsItem>
                  <MDBTabsLink
                    onClick={() => switchTab('1')}
                    active={activeItem === '1'}
                  >
                    Search by Id
                  </MDBTabsLink>
                </MDBTabsItem>
                <MDBTabsItem>
                  <MDBTabsLink
                    onClick={() => switchTab('2')}
                    active={activeItem === '2'}
                  >
                    Search by Username
                  </MDBTabsLink>
                </MDBTabsItem>
              </MDBTabs>

              <MDBTabsContent>
                <MDBTabsPane show={activeItem === '1'}>
                  <MDBInput
                    onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
                      handleChange(e)
                    }
                    value={query}
                    aria-label="Search ID"
                    label="Search ID"
                    icon="hashtag"
                    type="number"
                    error="wrong"
                    success="right"
                  />
                </MDBTabsPane>
                <MDBTabsPane show={activeItem === '2'}>
                  <MDBInput
                    onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
                      handleChange(e)
                    }
                    value={query}
                    aria-label="Search Username"
                    label="Search Username"
                    icon="user"
                    type="text"
                    error="wrong"
                    success="right"
                  />
                </MDBTabsPane>
                <FoundPlayersList
                  foundPlayers={selectedPlayer ?? foundPlayers}
                  handleSelect={handleSelect}
                />
              </MDBTabsContent>
            </MDBModalBody>

            {/* /// Show Buttons OR Status bar /// */}
            {!showSpinner && (
              <MDBModalFooter>
                {beeToAdd !== undefined && !userAlreadyAdded && !thatIsYou && (
                  <MDBBtn color="success" onClick={handleSave}>
                    Add {beeToAdd.playerName}
                  </MDBBtn>
                )}
                {beeToAdd === undefined && selectedPlayer !== undefined && (
                  <Spinner />
                )}
                {thatIsYou && (
                  <Alert
                    text={
                      'You found yourself! What a noble goal in life, but not helping you on Beesaber ;)'
                    }
                    type={'primary'}
                  />
                )}
                <MDBBtn color="danger" onClick={toggleModal}>
                  Close
                </MDBBtn>
              </MDBModalFooter>
            )}
            {showSpinner && (
              <MDBModalFooter>
                <NeonText glow as={'h2'} titleColor={'blue'}>
                  Searching...
                </NeonText>
                <Spinner />
              </MDBModalFooter>
            )}
          </ModalContent>
        </MDBModalDialog>
      </MDBModal>
    </MDBContainer>
  );
};

export default AddBeeModal;
