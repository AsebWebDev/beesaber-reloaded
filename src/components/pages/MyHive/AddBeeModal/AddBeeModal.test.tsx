import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import * as getAllScores from '@/api/services/helper/getAllScores';
import { initialState as store } from '@/store/store';

import AddBeeModal from './AddBeeModal';
import * as useQueryForPlayers from './useQueryForPlayers';

import type { ScoreData } from '@/sharedTypes';
import type { PlayerInfo } from '@/sharedTypes/ScoreSaberUserInfo';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('AddBeeModal', () => {
  let spy: jest.SpyInstance<
    {
      foundPlayers: PlayerInfo[] | null;
      showSpinner: boolean;
      thatIsYou: boolean;
      userAlreadyAdded: boolean;
    },
    [{ query: string; searchBy: 'id' | 'name' }]
  >;

  let getAllScoresSpy: jest.SpyInstance<
    Promise<ScoreData>,
    [id: string, threshold?: number | undefined]
  >;

  beforeEach(() => {
    spy = jest.spyOn(useQueryForPlayers, 'default');
    getAllScoresSpy = jest.spyOn(getAllScores, 'default');
    jest.useFakeTimers();
  });

  afterEach(() => {
    spy.mockRestore();
    getAllScoresSpy.mockRestore();
    spy.mockReset();
    getAllScoresSpy.mockReset();
    jest.useRealTimers();
  });

  it('should match snapshot', () => {
    spy.mockReturnValueOnce({
      foundPlayers: null,
      showSpinner: false,
      thatIsYou: false,
      userAlreadyAdded: false,
    });
    const { container } = render(
      <Provider store={mockStore(store)}>
        <AddBeeModal toggleModal={jest.fn()} />
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should show Spinner, when showSpinner is true', () => {
    spy.mockReturnValueOnce({
      foundPlayers: null,
      showSpinner: true,
      thatIsYou: false,
      userAlreadyAdded: false,
    });

    render(
      <Provider store={mockStore(store)}>
        <AddBeeModal toggleModal={jest.fn()} />
      </Provider>
    );

    const spinner = screen.getByRole('heading', { name: 'Searching...' });
    const buttons = screen.queryByText('Close');

    expect(buttons).not.toBeInTheDocument();
    expect(spinner).toBeInTheDocument();
  });

  // FIXME: when MSW is running for tests, to properly test useGetUserDataQuery hook

  // it('should show Buttons, when showSpinner is false', () => {
  //   spy.mockReturnValueOnce({
  //     foundPlayers: null,
  //     showSpinner: false,
  //     thatIsYou: false,
  //     userAlreadyAdded: false,
  //   });

  //   render(
  //     <Provider store={mockStore(store)}>
  //       <AddBeeModal toggleModal={jest.fn()} />
  //     </Provider>
  //   );

  //   const spinner = screen.queryByText('Searching...');
  //   const buttons = screen.getByRole('button', { name: 'Close' });

  //   expect(buttons).toBeInTheDocument();
  //   expect(spinner).not.toBeInTheDocument();
  // });

  // it('should call toggleModal when close button is clicked', () => {
  //   const mockToggleModal = jest.fn();

  //   spy.mockReturnValue({
  //     foundPlayers: null,
  //     showSpinner: false,
  //     thatIsYou: false,
  //     userAlreadyAdded: false,
  //   });
  //   render(
  //     <Provider store={mockStore(store)}>
  //       <AddBeeModal toggleModal={mockToggleModal} />
  //     </Provider>
  //   );
  //   const closeButton = screen.getByRole('button');
  //   act(() => userEvent.click(closeButton));
  //   expect(mockToggleModal).toBeCalledTimes(1);
  // });

  // it('should show feedback when thatIsYou is true', () => {
  //   spy.mockReturnValue({
  //     foundPlayers: [exampleSSUserInfo.playerInfo],
  //     showSpinner: false,
  //     thatIsYou: true,
  //     userAlreadyAdded: false,
  //   });
  //   getAllScoresSpy.mockResolvedValue(exampleScoreData)
  //   render(
  //     <Provider store={mockStore(store)}>
  //       <AddBeeModal toggleModal={jest.fn()} />
  //     </Provider>
  //   );
  //   const feedback = screen.getByText('That is you')
  //   expect(feedback).toBeInTheDocument();
  // });

  // it('should show "add" button, when valid player is selected', () => {
  //   spy.mockReturnValue({
  //     foundPlayers: [exampleSSUserInfo.playerInfo],
  //     showSpinner: false,
  //     thatIsYou: true,
  //     userAlreadyAdded: false,
  //   });
  //   getAllScoresSpy.mockResolvedValue(exampleScoreData)
  //   render(
  //     <Provider store={mockStore(store)}>
  //       <AddBeeModal toggleModal={jest.fn()} />
  //     </Provider>
  //   );
  //   const addButton = screen.getByRole(`Add ${exampleSSUserInfo.playerInfo.playerName}`);
  //   expect(addButton).toBeInTheDocument();
  // });
  // it('should show players, when multiple players are found', () =>{
  //   spy.mockReturnValue({
  //     foundPlayers: examplePlayerInfos,
  //     showSpinner: false,
  //     thatIsYou: false,
  //     userAlreadyAdded: false,
  //   });

  //   render(
  //     <Provider store={mockStore(store)}>
  //       <AddBeeModal toggleModal={jest.fn()} />
  //     </Provider>
  //   );
  //   const playersList = screen.findByTestId('found-players-list')
  //   expect(playersList).toBeInTheDocument();
  // })

  it.todo('should match snapshot with no found player');
  it.todo('should match snapshot with one found player');
  it.todo('should match snapshot with multiple found players');
  it.todo('should show selected player, when a player is selected');
  it.todo('should not show "add" button, when no player is selected');
  it.todo(
    'should not show "add" button, when selected player is already added'
  );
  it.todo('should not show "add" button, when selected player is you');

  it.todo(
    'should show Button Spinner, when player is selected, but data is still loading'
  );
  it.todo('should show Footer Spinner, when search is in progress');
  it.todo('should call handleSave when save button is clicked');
  it.todo('should call switchTab when tab is clicked');
  it.todo('should reset query when tab is switched');
  it.todo('should reset query after handleSave is done');
  it.todo('should select player, when handleSelect is called');
  it.todo('should build a correct bee object, when player is selected');
});
