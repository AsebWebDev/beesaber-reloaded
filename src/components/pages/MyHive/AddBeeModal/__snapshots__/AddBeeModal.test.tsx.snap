// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBeeModal should match snapshot 1`] = `
.c0 {
  background-color: #162530;
  color: rgba(255,255,255,1);
  margin-top: 10rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

<div
  class="container"
>
  <div
    class="modal fade show"
    style="display: block;"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content c0"
      >
        <div
          class="modal-header"
        >
          Add a new Bee
        </div>
        <div
          class="modal-body"
        >
          <ul
            class="nav nav-tabs mb-3"
          >
            <li
              class="nav-item"
              role="presentation"
              style="cursor: pointer;"
            >
              <a
                class="nav-link active"
              >
                Search by Id
              </a>
            </li>
            <li
              class="nav-item"
              role="presentation"
              style="cursor: pointer;"
            >
              <a
                class="nav-link"
              >
                Search by Username
              </a>
            </li>
          </ul>
          <div
            class="tab-content"
          >
            <div
              class="tab-pane fade active"
              role="tabpanel"
            >
              <div
                class="form-outline"
              >
                <input
                  class="form-control"
                  error="wrong"
                  icon="hashtag"
                  success="right"
                  type="number"
                  value=""
                />
                <label
                  class="form-label"
                >
                  Search ID
                </label>
                <div
                  class="form-notch"
                >
                  <div
                    class="form-notch-leading"
                  />
                  <div
                    class="form-notch-middle"
                    style="width: 0px;"
                  />
                  <div
                    class="form-notch-trailing"
                  />
                </div>
              </div>
            </div>
            <div
              class="tab-pane fade"
              role="tabpanel"
            >
              <div
                class="form-outline"
              >
                <input
                  class="form-control"
                  error="wrong"
                  icon="user"
                  success="right"
                  type="text"
                  value=""
                />
                <label
                  class="form-label"
                >
                  Search Username
                </label>
                <div
                  class="form-notch"
                >
                  <div
                    class="form-notch-leading"
                  />
                  <div
                    class="form-notch-middle"
                    style="width: 0px;"
                  />
                  <div
                    class="form-notch-trailing"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="ripple ripple-surface btn btn-danger"
            role="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
